internal/mesh/internal/types/upstreams.go

```diff
@@ -28,6 +32,82 @@ func RegisterUpstreams(r resource.Registry) {
 		Type:     UpstreamsV1Alpha1Type,
 		Proto:    &pbmesh.Upstreams{},
 		Scope:    resource.ScopeNamespace,
-		Validate: nil,
+		Mutate:   MutateUpstreams,
+		Validate: ValidateUpstreams,
 	})
 }
+
+func MutateUpstreams(res *pbresource.Resource) error {
+	var route pbmesh.Upstreams
+
+	if err := res.Data.UnmarshalTo(&route); err != nil {
+		return resource.NewErrDataParse(&route, err)
+	}
+
+	changed := false
+
+	for _, dest := range route.Upstreams {
+		if dest.DestinationRef.Tenancy != nil && !isLocalPeer(dest.DestinationRef.Tenancy.PeerName) {
+			// TODO(peering/v2): remove this bypass when we know what to do with
+			// non-local peer references.
+			continue
+		}
+		orig := proto.Clone(dest.DestinationRef).(*pbresource.Reference)
+		resource.DefaultReferenceTenancy(
+			dest.DestinationRef,
+			res.Id.GetTenancy(),
+			resource.DefaultNamespacedTenancy(), // Services are all namespace scoped.
+		)
+
+		if !proto.Equal(orig, dest.DestinationRef) {
+			changed = true
+		}
+	}
+
+	if !changed {
+		return nil
+	}
+
+	return res.Data.MarshalFrom(&route)
+}
+
+func isLocalPeer(p string) bool {
+	return p == "local" || p == ""
+}
+
+func ValidateUpstreams(res *pbresource.Resource) error {
+	var route pbmesh.Upstreams
+
+	if err := res.Data.UnmarshalTo(&route); err != nil {
+		return resource.NewErrDataParse(&route, err)
+	}
+
+	var merr error
+
+	for i, dest := range route.Upstreams {
+		wrapDestErr := func(err error) error {
+			return resource.ErrInvalidListElement{
+				Name:    "upstreams",
+				Index:   i,
+				Wrapped: err,
+			}
+		}
+
+		wrapRefErr := func(err error) error {
+			return wrapDestErr(resource.ErrInvalidField{
+				Name:    "destination_ref",
+				Wrapped: err,
+			})
+		}
+
+		if refErr := catalog.ValidateLocalServiceRefNoSection(dest.DestinationRef, wrapRefErr); refErr != nil {
+			merr = multierror.Append(merr, refErr)
+		}
+
+		// TODO(v2): validate port name using catalog validator
```

Comment:
I didn't want to balloon this PR to fully validate this stuff.

Reply:
Do you mind adding a todo for validating `ListenAddr` too?